File: packages/framework-adaptor/mcp-nethttp/go/middleware.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// net/http middleware that delegates MCP auth to transport-mcp.

package mcpnethttp

import (
	"encoding/json"
	"log"
	"net/http"

	transportmcp "github.com/garudex-labs/caracal/transport-mcp"
)

// Options configures the auth middleware.
type Options = transportmcp.Options

type errBody struct {
	Error            string `json:"error"`
	ErrorDescription string `json:"error_description"`
}

// Middleware returns a net/http middleware that validates Caracal JWTs.
func Middleware(opts Options) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			token, _ := transportmcp.ExtractBearer(r.Header.Get("Authorization"))
			_, authErr := transportmcp.Authenticate(token, opts)
			if authErr != nil {
				status, code := mapError(authErr.Code)
				writeErr(w, status, code, authErr.Description)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

func mapError(code transportmcp.ErrorCode) (int, string) {
	if code == transportmcp.ErrInsufficientScope {
		return http.StatusForbidden, "insufficient_scope"
	}
	return http.StatusUnauthorized, "invalid_token"
}

func writeErr(w http.ResponseWriter, status int, code, desc string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(errBody{Error: code, ErrorDescription: desc}); err != nil {
		log.Printf("mcp-nethttp: failed to encode error response: %v", err)
	}
}

File: packages/connectors/fastmcp/ts/src/middleware.ts
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// FastMCP token verifier that delegates to @caracalai/transport-mcp.

import { authenticate, extractBearer } from '@caracalai/transport-mcp'
import type { AuthDeps, AuthErrorCode } from '@caracalai/transport-mcp'

export type FastMcpAuthOptions = AuthDeps

export interface FastMcpContext {
  sub: string
  zoneId: string
  scope: string
}

export class FastMcpAuthError extends Error {
  readonly code: AuthErrorCode
  constructor(code: AuthErrorCode, description: string) {
    super(description)
    this.name = 'FastMcpAuthError'
    this.code = code
  }
}

export async function verifyFastMcpToken(
  token: string,
  opts: FastMcpAuthOptions,
): Promise<FastMcpContext> {
  const result = await authenticate(token, opts)
  if (!result.ok) throw new FastMcpAuthError(result.error.code, result.error.description)
  const claims = result.principal
  return { sub: claims.sub, zoneId: claims.zoneId, scope: claims.scope }
}

export { extractBearer }
